File: backend/app/schemas/todo.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class TodoBase(BaseModel):
    """Shared fields for Todo schemas."""

    title: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    priority: Literal["low", "medium", "high"] = "medium"


class TodoCreate(TodoBase):
    """Schema for creating a new Todo."""


class TodoUpdate(BaseModel):
    """Schema for partially updating a Todo (all fields optional)."""

    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    completed: bool | None = None
    priority: Literal["low", "medium", "high"] | None = None


class TodoResponse(TodoBase):
    """Schema for Todo API responses."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    completed: bool
    created_at: datetime
    updated_at: datetime | None = None


class TodoListResponse(BaseModel):
    """Schema for paginated Todo list responses."""

    items: list[TodoResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/todo.ts
export type Priority = 'low' | 'medium' | 'high';

export interface Todo {
  id: string;
  title: string;
  description: string | null;
  completed: boolean;
  priority: Priority;
  created_at: string;
  updated_at: string | null;
}

export interface TodoCreate {
  title: string;
  description?: string | null;
  priority?: Priority;
}

export interface TodoUpdate {
  title?: string;
  description?: string | null;
  completed?: boolean;
  priority?: Priority;
}

export interface TodoListResponse {
  items: Todo[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ApiMeta {
  version: string;
}

export interface ApiErrorDetail {
  code: string;
  message: string;
}

export interface ApiEnvelope<T> {
  data: T | null;
  meta: ApiMeta;
  error: ApiErrorDetail | null;
}

export interface TodoFilters {
  search?: string;
  status?: 'all' | 'active' | 'completed';
  priority?: 'all' | Priority;
}
